manpages/man/tiledb_export_csv.1
.TH tiledb_export_csv 1 "27 June 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_export_csv - exports the (binary) data of an array into a CSV file

.SH SYNOPSIS
.B tiledb_export_csv 
.BI "-w " "directory " "-A " "array_name " "-f " "filename" 
[
.BI "-a " "attribute_name_1,..."
]
[
.BI "-d " "dim_name_1,..."
]
[
.BI "-m " "mode"
]

.SH DESCRIPTION
.B tiledb_export_csv
exports the (binary) data of an array into a CSV file. The documentation on
the exported CSV format can be found in program \fBtiledb_load_csv\fR.
On error, it prints a message on stderr and returns an error code 
(see \fBEXIT STATUS\fR). 

The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
.TE

The user may also specify a subset of attributes and dimensions from the
array to export. Also, he/she may specify any order of attributes and 
dimensions, as well as multiplicities. 
 
In case the user wishes to completely hide the dimensions or the attributes,
he/she should include \fB__hide\fR as the only name in the dimension 
or attribute name list.

Finally, the user may specify whether the array cells should be exported
in the stored order or in \fBreverse\fR.
 
.SH OPTIONS
.TP
.BI "-w" " directory" "\fR, " \fB --workspace=\fIdirectory\fR  
The path to the workspace folder where the TileDB are stored. The path
must exist, and the caller must have read and write permissions on it.

.TP
.BI "-A" " array_name" "\fR, " \fB --array-name=\fIarray_name\fR  
The name of the array whose data will be exported. The array
must be already defined. 

.TP
.BI "-f" " filename" "\fR, " \fB --filename=\fIfilename\fR  
The name of the CSV file. 

.TP
.BI "-a" " attribute_name_1,..." "\fR, " \
\fB --attribute-names=\fIattribute_name_1,...\fR  
The attributes to be exported. Any order and multiplicities are allowed. To
export \fBno\fR attribute, include \fB__hide\fR instead of a name list. To
export \fBall\fR attributes, omit the option.

.TP
.BI "-d" " dim_name_1,..." "\fR, " \
\fB --dim-names=\fIdim_name_1,...\fR  
The coordinates to be exported. Any order and multiplicities are allowed. To
export \fBno\fR coordinates, include \fB__hide\fR instead of a name list. To
export \fBall\fR coordinates, omit the option.

.TP
.BI "-m" " mode" "\fR, " \fB --mode=\fImode\fR  
There are two modes: \fBnormal\fR and \fBreverse\fR. In the former, the cells
are exported in the same order they are stored (i.e., the native order specified
in the array schema). In the latter, the cells are exported in the reverse order
from the one they are stored. If the option is omitted, the default is 
\fBnormal\fR.

.SH EXAMPLES
Assuming that the array has dimensions \fIdim1\fR,\fIdim2\fR and attributes 
\fIattr1\fR,\fIattr2\fR,\fIattr3\fR
.TP
\fBtiledb_export_csv\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv

Exports into \fImy_array.csv\fR all the cells of array \fImy_array\fR in the order they are 
stored, including all the coordinates and attributes.

.TP
\fBtiledb_export_csv\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv \\
    \fB-d \fRdim1 \\
    \fB-a \fRattr1,attr2

Exports int \fImy_array.csv\fR all the cells of \fImy_array\fR in the order they are stored,
including only the coordinates on the \fIdim1\fR dimension and the values
on the \fIattr1\fR,\fIattr2\fR attributes.

.TP
\fBtiledb_export_csv\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv \\
    \fB-d \fRdim1 \\
    \fB-a \fRattr1,attr2 \\
    \fB-m \fRreverse

Same as above, but the cells are exported in the reverse order.

.TP
\fBtiledb_export_csv\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv \\
    \fB-d \fR__hide \\
    \fB-a \fRattr1,attr2 

 Same as the second example, but no coordinate is exported.

.TP
\fBtiledb_export_csv\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv \\
    \fB-d \fRdim1 \\
    \fB-a \fR__hide

Same as the second example, but no attribute value is exported.

.TP
\fBtiledb_export_csv\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.csv \\
    \fB-d \fRdim1 \\
    \fB-a \fRattr1,attr2,attr1

Same as the second example, but now the \fIattr1\fR values are shown twice
(once before those of \fIattr2\fR and once after).

.SH EXIT STATUS
.TP 
.B 0
Successful execution
.TP 
.B -1
Parser error
.TP 
.B -2
Undefined array
.TP 
.B -3
File operation failed
.TP 
.B -7
TileDB initialization failed
.TP 
.B -8
TileDB finalization failed
.TP 
.B -14
Failed to open array
.TP 
.B -15
Failed to close array

.SH SEE ALSO
.BR "tiledb_load_csv" "(1) " "tiledb_generate_data" "(1) "

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_export_csv.cc
/**
 * @file   tiledb_export_csv.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2014 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_export_csv".
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include <getopt.h>
#include <iostream>
#include <string>

/** 
 * Parses the command options. It returns 0 on success. On error, it prints
 * a message on stderr and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& array_name, 
    std::string& workspace,
    std::string& filename,
    std::vector<std::string>& dim_names,
    std::vector<std::string>& attribute_names,
    bool& reverse) {
  // Initialization
  array_name = "";
  workspace = "";
  filename = "";
  std::string attribute_names_str("");
  std::string dim_names_str("");
  std::string mode("");
  dim_names.clear();
  attribute_names.clear();

  // Define long options
  static struct option long_options[] = 
  {
    {"attribute-names",1,0,'a'},
    {"array-name",1,0,'A'},
    {"dim-names",1,0,'d'},
    {"filename",1,0,'f'},
    {"mode",1,0,'m'},
    {"workspace",1,0,'w'},
    {0,0,0,0},
  };
  // Define short options
  const char* short_options = "a:A:d:f:m:w:";
  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'a':
        if(attribute_names_str != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one attribute name lists provided.\n";
          return -1;
        }
        attribute_names_str = optarg;
        break;
      case 'A':
        if(array_name != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one array names provided.\n";
          return -1;
        }
        array_name = optarg;
        break;
      case 'd':
        if(dim_names_str != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one dimension name lists provided.\n";
          return -1;
        }
        dim_names_str = optarg;
        break;
      case 'f':
        if(filename != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one CSV file names provided.\n";
          return -1;
        }
        filename = optarg;
        break;
      case 'm':
        if(mode != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one modes provided.\n";
          return -1;
        }
        mode = optarg;
        break;
      case 'w':
        if(workspace != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one workspaces provided.\n";
          return -1;
        }
        workspace = optarg;
        break; 
      default:
        return -1;
    }
  }

  // Check number of arguments
  if((argc-1) != 2*option_num) {
    std::cerr << ERROR_MSG_HEADER << " Arguments-options mismatch.\n";
    return -1;
  }
  // Check if correct arguments have been ginen
  if(array_name == "") {
    std::cerr << ERROR_MSG_HEADER << " Array name not provided.\n";
    return -1;
  }
  if(workspace == "") {
    std::cerr << ERROR_MSG_HEADER << " Workspace not provided.\n";
    return -1;
  }
  if(filename == "") {
    std::cerr << ERROR_MSG_HEADER << " CSV file name not provided.\n";
    return -1;
  }

  // Check number of arguments
  // ----- array name
  CSVLine temp;
  temp << array_name;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one array names provided.\n";
    return -1;
  }
  // ----- workspace
  temp.clear();
  temp << workspace;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one workspaces provided.\n";
    return -1;
  }
  // ----- filename
  temp.clear();
  temp << filename;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one CSV file names provided.\n";
    return -1;
  }
  // ----- mode
  temp.clear();
  temp << mode;
  if(temp.val_num() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one modes provided.\n";
    return -1;
  }

  // Check mode
  if(mode == "" || mode == "normal") {
    reverse = false;
  } else if(mode == "reverse") { 
    reverse = true; 
  } else {
    std::cerr << ERROR_MSG_HEADER << " Unknown mode.\n";
    return -1;
  }

  // Get dimension names
  if(dim_names_str != "")
    dim_names = CSVLine(&dim_names_str[0]).values_str_vec();

  // Get attribute names
  if(attribute_names_str != "")
    attribute_names = CSVLine(&attribute_names_str[0]).values_str_vec();

  return 0;
}

int main(int argc, char** argv) {
  // Arguments
  std::string array_name;
  std::string workspace;
  std::string filename;
  std::vector<std::string> dim_names, attribute_names;
  bool reverse;
  const char** dim_names_c_str;
  const char** attribute_names_c_str;
  int dim_names_num, attribute_names_num;

  // Stores the return code of the various functions below
  int rc; 
 
  // Parse command line
  rc = parse_options(
           argc, argv, array_name, workspace, filename,
           dim_names, attribute_names, reverse);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to parse the command line.\n";
    return TILEDB_EPARSE;
  }

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  rc = tiledb_ctx_init(workspace.c_str(), tiledb_ctx);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to initialize TileDB.\n";
    return TILEDB_EINIT;
  }

  // Populate dim_names_c_str
  dim_names_num = int(dim_names.size());
  dim_names_c_str = new const char*[dim_names_num];
  for(int i=0; i<dim_names_num; ++i) 
    dim_names_c_str[i] = dim_names[i].c_str();
  // Populate attribute_names_c_str
  attribute_names_num = int(attribute_names.size());
  attribute_names_c_str = new const char*[attribute_names_num];
  for(int i=0; i<attribute_names_num; ++i) 
    attribute_names_c_str[i] = attribute_names[i].c_str();

  // Export to CSV
  rc = tiledb_export_csv(
           tiledb_ctx, array_name.c_str(), filename.c_str(),
           dim_names_c_str, dim_names_num, 
           attribute_names_c_str, attribute_names_num,
           reverse);
  if(rc) { 
    delete [] dim_names_c_str;
    delete [] attribute_names_c_str;
    return rc;
  }

  // Finalize TileDB
  rc = tiledb_ctx_finalize(tiledb_ctx);
  if(rc) {
    delete [] dim_names_c_str;
    delete [] attribute_names_c_str;
    std::cerr << ERROR_MSG_HEADER << " Failed to finalize TileDB.\n";
    return TILEDB_EFIN;
  }

  std::cout << MSG_HEADER << " Program executed successfully!\n";

  // Clean up  
  delete [] dim_names_c_str;
  delete [] attribute_names_c_str;

  return 0;
}
